src/main/java/io/ifar/archive/core/ArchiveListener.java

```diff
@@ -48,11 +56,43 @@ public ArchiveListener(AmazonS3Client s3Client, String seedBrokers, String workU
         this.executor = Executors.newScheduledThreadPool(maxNumParallelWorkers);
         this.metrics = metrics;
         this.topicConfiguration = topicConfiguration;
+        this.errorLimitConfiguration = errorLimitConfiguration;
         this.defaultBucket = defaultBucket;
 
         numWorkUnitsGauge = metrics.register("archiveListener-workUnits", new IntegerGauge());
     }
 
+    public void setCluster(ArchiveCluster archiveCluster) {
+        this.archiveCluster = archiveCluster;
+    }
+
+    public void batchFailure(String workUnit) {
+        if (!stopFlag.get()) {
+            int errors = failedBatches.getAndIncrement()+1;
+            LOG.warn("Encountered error in archive worker {} ({} error(s) total)", workUnit, errors);
+            int successful = successfulBatches.get();
+
+            if (errors >= errorLimitConfiguration.getTriesBeforeEnableErrorLimit()) {
+
+                double limit = errorLimitConfiguration.getErrorLimitPercent() / 100.0;
+
+                if ( (1.0*errors) / (1.0*successful) >= limit ) {
```

Comment:
On this logic: if system has been running successfully for a while, doesn't this mean that number of errors needed becomes rather high too? Would it perhaps be better to use a statistics package like coda metrics, to check for error rates for, say, past 5 minutes?